### ISSUE TITLE
Add an online playground and initial project structure

### ISSUE THREAD

**Issue by adelhult** [modmark-org/modmark#5]
For us to get started on the project and easily visualize the inner workings of the system it would be a great idea to add a web based playground (similar to [djot playground](https://djot.net/playground/)) that gets built automatically on every commit to the main branch. 
 
**Describe the solution you'd like**
Rust projects are often structured into multiple smaller disjoint library crates that then are sown together in another crate that imports them and provides an interface, this could for instance be a terminal app or bindings to an editor. 

For Modmark we could structure our repo into these main directories / crates:
```
parser (Takes text as input and produces a document tree)
compiler (Evaluates the document tree and hosts the wasm runtime. Might want to call this runtime/core instead...)
cli (Terminal tool that users will interact with)
playground (Allows us to inspect the document tree and play around with the parser and compiler)
```  

We should of course first and foremost focus on research and planning these first few weeks. But if we get this infrastructure set up we would have a good shared framework that we all can use to play around with our ideas.

**Describe alternatives you've considered**
We could just create a very simple terminal tool. But I think a website works better when we later down the line need to debug large large document trees.

**Comment by adelhult** [comment]
I'll maybe throw together for this during the week-end, but would love to here if you thing it's a good idea or have another structure in mind.

**Comment by CMDJojo** [comment]
Sounds good! I think it is a good idea to structure the repo into different crates. It makes sense to have one crate that is for the `cli` and one for the `playground` since they are two different wrappers around Modmark which has little to do with the language itself.

However, I think that we should have another structure for the other crates: a `core`/`runtime`/`modmark`-crate would be useful, which will be the one scheduling running the modules and such, keeping track of state and so on, but I do think we should have a `moduleloader`/`modules`/`modmgr`-module which finds, loads and keeps track of loaded modules. Since the data we will pass from/to a module will be an AST which will be transformed to another AST, with both of them being serializable (to json or something), there is nothing hindering us from letting `parser` have the same interface. Regardless if we will treat the `parser` as something configurable/modular like any other module, we can use the normal module interface for interacting with it in the first place. If we would decide that we should have special rule for it (which we most certainly will have in some form), we can just restrict the access compared to the other modules, which we might have to do for some other modules as well. My point is that I think the parser should be living along side the other standard modules and I don't see the need to create a special crate just for the parser itself; we will probably need a lot of standard modules (on the order of 20 is my best guess), so it would make sense to store the standard modules within whatever crate is managing them (probably the `modules` crate).

Being able to test the parser is very vital, I can see that, and that is why I would suggest starting with `playground`, implementing the `core` as just "reading the text, sending it to `modmgr` which in turn just runs the `parser` module, then the core would serialize the resulting `AST` and return the result.

Does this make sense? Does it sound reasonable? Would love to hear any other thoughts as well!

**Comment by CMDJojo** [comment]
Additionally, we will probably need a sort of "standard library" for the modules to interact with, for creating blobs or using some other system interfacing function which we probably want to restrict the access to, and I am uncertain to where that library should be located. I think it is reasonable to include it in the `moduleloader`/`modules`/`modmgr` crate as a start, and if we feel like it is taking up too much space, we might split it out to another crate later on.

So, my proposed structure is:
```
cli (terminal tool that users will interact with)
playground (allows us to inspect the document tree and play around with the parser and compiler)
runtime (evaluates the document tree and hosts the wasm runtime. mostly a scheduler and keeping track of current state)
modmanager (a package mgr, hosts the built-in modules, including parser, loads modules etc)
modstdlib (the "standard library" for modules to interact with, might be included in modmanager if not too large)
```

The standard modules might be hosted in individual crates, or in one large crate separate from the package manager later on, if we see fit.

**Comment by adelhult** [comment]
Great input!

I would argue that `module_loader` could get away with being a normal Rust sub-module in the `core` crate and that we might not need a totally separate crate since managing modules probably will be very tied to the core functionality. But, I guess this does not matter much right now anyhow. I'm totally fine with adding that if we see fit once we have started working on the implementation of the module system.

I totally agree with you on that we should have top level directory for  `standard_modules` that contains all of the "standard library" modules such as math, figures etc. Especially since we might want to compile them to WASM files during the build step and let them use the same interface as other third party modules. All standard models might not even need to written in Rust if we opt for this structure.

I'm a bit hesitant on putting the parser amongst those other modules though (at least now during the start of the project). For the following reasons: 

1. If we wish to support other languages other then Rust, being able to build the parse disjoint from everything else helps a lot, we can just compile the rust implementation and add bindings.
2. Having the parser live in a separate crate allows us to work in parallel without everyone needing to be concerned about the inner workings of the module system at the start of the project.
3. We can always move the parser later once we see how everything should fit together. Another option is to still have a parsing modules amongst the other standard modules that just imports the top-level parser crate. This way can still have the benefits of your idea if that proves to be the way to go later down the line.

**Comment by CMDJojo** [comment]
Hi Eli!

Nice idea with having the module loader as a sub-create, it seems reasonable. I didn't realize that was an option due to my lack of experience with Rust and Cargo.

I also see your points with having the parser separate, however I feel that most of them aren't so problematic.

1. The idea of having the parser disjoint from everything else for interop with other languages is reasonable, however I would say that it wouldn't be very hard for someone wanting to use the same parser to just copy it out of the module and into their own project/file.
2. I don't think it would hinder us from working in parallel at all; we can, for the first test-versions, just hard-code the module loader/core to just run the parser and spit out the result, as I suggested earlier. This will mean that progression of the parser may be done without knowledge of the module loader, and the module loader may be worked on at the same time (as long as it at some point runs the parser) without interrupting the parser. The parser is just text-in text-out, the module system doesn't have to be near fully developed when testing
3. I would say that the main job of the parser is accepting text input and returning text output, maybe adding some stuff to the document state as well. This is identical to how modules will work. Additionally, we might want to be able to configure the parser from within the document, like to enable/disable soft-wraps etc, similarly to as how we may want to configure other modules for other stuff. Having them both use the same interface would keep the syntax more consistent and ensure that developments done to one system benefits the other as well. Separating them early in hope of joining them later on might split development efforts; joining them might lead to developments of the parser that would better benefit other modules, such as functions for the standard library

It is hard to know what the best solution is, and one compromise might be your third option with making a module which is just a wrapper for the parser living in a separate crate. This might, however, make the parser not benefit as much of the standard library grows. However, if you are the one actually making and pulling the first draft, you have the final call!

**Comment by adelhult** [comment]
Yeah, sounds reasonable. I think I will opt for the easier structure for now once making a first draft, then we can restructure it if we see fit once we get some more of the parts sketched out : )
